### PR TITLE
Removed const from algorithms

### PR DESCRIPTION
**Pull request by dcbdan**
This pull request is an extension of #1556. 
- All consts related to parallel algorithms (not segmented parallel algorithms, though) have been removed so that non-const execution policies can be used. This includes all of the algorithm implementations in `hpx/parallel/algorithms` and tests in `tests/unit/parallel/algorithms`.
- The template specialization was changed for each of the partitioners so that task execution policies work with the .on syntax. It was previously ambiguous because the Tag typename was also free. The added tests inclusive_scan_executors, is_sorted_executors, uninitialized_copy_executors are meant to test the changes on the scan_partitioner, partitioner and partitioner_with_cleanup respectively. (similar to the changes to the foreach partitioner, #1556)

[Review on Reviewable](https://reviewable.io/reviews/stellar-group/hpx/1567)

### CODE REVIEW

**Comment by sithhell** [comment]
The commit message should be more descriptive about what has changed. From the title of the PR it's not very clear where `const` has been removed.

File: hpx/parallel/util/partitioner.hpp [hunk 1]
```diff
@@ -429,10 +435,28 @@ namespace hpx { namespace parallel { namespace util
...
+          : partitioner<parallel_task_execution_policy, R, Result,
+                parallel::traits::auto_partitioner_tag>
+        {};
+
+        template <typename Executor, typename R, typename Result>
+        struct partitioner<
+                parallel_task_execution_policy_shim<Executor>, R, Result,
+                parallel::traits::default_partitioner_tag>
+          : partitioner<parallel_task_execution_policy, R, Result,
+                parallel::traits::static_partitioner_tag>
```
Comment (sithhell) [inline hpx/parallel/util/partitioner.hpp]: Why are those specializations needed?

File: hpx/parallel/util/partitioner_with_cleanup.hpp [hunk 2]
```diff
@@ -302,12 +309,31 @@ namespace hpx { namespace parallel { namespace util
...
 
+        template <typename Executor, typename R, typename Result>
+        struct partitioner_with_cleanup<
+                parallel_task_execution_policy_shim<Executor>, R, Result,
+                parallel::traits::default_partitioner_tag>
+          : partitioner_with_cleanup<parallel_task_execution_policy, R, Result,
+                parallel::traits::static_partitioner_tag>
+        {};
+
+
```
Comment (sithhell) [inline hpx/parallel/util/partitioner_with_cleanup.hpp]: Same as above.

File: hpx/parallel/util/scan_partitioner.hpp [hunk 3]
```diff
@@ -248,10 +254,28 @@ namespace hpx { namespace parallel { namespace util
...
+          : scan_partitioner<parallel_task_execution_policy, R, Result,
+                parallel::traits::auto_partitioner_tag>
+        {};
+
+        template <typename Executor, typename R, typename Result>
+        struct scan_partitioner<
+                parallel_task_execution_policy_shim<Executor>, R, Result,
+                parallel::traits::default_partitioner_tag>
+          : scan_partitioner<parallel_task_execution_policy, R, Result,
+                parallel::traits::static_partitioner_tag>
```
Comment (sithhell) [inline hpx/parallel/util/scan_partitioner.hpp]: Why do we need those all of sudden?

**Comment by dcbdan** [comment]
@sithhell: The pull request description was edited to clarify which consts and why the further template specifications was needed. Thanks for pointing out that my commit message was unclear--I didn't notice at the time. I don't know if that can be changed, though.

File: hpx/parallel/util/scan_partitioner.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/parallel/util/scan_partitioner.hpp]: Those are needed as otherwise our existing specializations would be ambiguous.

**Comment by hkaiser** [comment]
If there are no objections I will merge this later today.

**Pull request merged**